lib/Dialect/TritonGPU/IR/LinearLayoutConversions.cpp

```diff
@@ -877,20 +877,31 @@ SliceEncodingAttr::toLinearLayout(ArrayRef<int64_t> shape) const {
 std::optional<LinearLayout>
 toLinearLayout(ArrayRef<int64_t> shape, Attribute layout,
                std::optional<int32_t> elemBitWidth /*= std::nullopt*/) {
-  // Layouts are distributed or shared
+  CacheKey key{std::vector<int64_t>(shape.begin(), shape.end()), layout,
+               elemBitWidth};
+  auto *ctx = layout.getContext();
+  auto &llCache = ctx->getLoadedDialect<TritonGPUDialect>()->llCache;
+  auto result = llCache.get(key);
+  if (result.has_value()) {
+    return result;
+  }
+
+  // Layouts are distributed or shared in triton core
   if (auto distributed = dyn_cast<DistributedEncodingTrait>(layout)) {
-    return distributed.toLinearLayout(shape);
+    result = distributed.toLinearLayout(shape);
   } else if (auto shared = dyn_cast<SharedEncodingAttr>(layout)) {
     if (shared.getHasLeadingOffset()) {
       assert(elemBitWidth.has_value());
-      return sharedToLinearLayoutLeadingOffset(shape, shared, *elemBitWidth);
+      result = sharedToLinearLayoutLeadingOffset(shape, shared, *elemBitWidth);
     } else {
-      return sharedToLinearLayoutNoLeadingOffset(shape, shared);
+      result = sharedToLinearLayoutNoLeadingOffset(shape, shared);
     }
   }
 
-  // Third party layouts
-  return std::nullopt;
+  if (result.has_value()) {
+    llCache.set(std::move(key), *result);
```

Comment:
```suggestion
    llCache.set(std::move(key), std::move(*result));
```

Reply:
This is what I initially wrote... and it's a bug! we are returning `result` a couple lines down